### PR TITLE
merlin overrides use interval tree instead of list

### PR DESCRIPTION
**Pull request by maxmwang**
Implements interval tree and refactors overrides to use this new data structure. The chosen interval tree design is based off of [this description](https://en.wikipedia.org/wiki/Interval_tree#Centered_interval_tree). 

## Testing 

Added unit tests for invalid inputs and common happy path cases. 

Testing story can possibly be improved with more edge case testing.

### CODE REVIEW

File: tests/test-dirs/overrides.t [hunk 1]
```diff
@@ -373,9 +373,17 @@ Attribute location should not affect functionality.
   > [@@@merlin.document
   >   [ { location =
   >         { loc_start =
-  >             { pos_fname = "attribute-at-top.ml"; pos_lnum = 31; pos_bol = 867; pos_cnum = 880 }
```
[the remark this answers is not in the document]
Comment (maxmwang) [reply tests/test-dirs/overrides.t:376]: The test started failing with the interval tree for some reason, so I regenerated the attributes which caused them to be reformatted. 

It is strange that it used to pass, but I suspect its because range endpoints were treated inclusively, and then in the code of this review, they are treated exclusively.

File: tests/test-units/analysis/overrides_interval_tree_test.ml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      let payload = Overrides_interval_tree.find tree 0 in
+      check (option string) "should be equal" expected payload)
+
+let cases =
+  ( "overrides-interval-tree",
+    [ test_of_alist_exn;
+      test_invalid_interval;
+      test_find ~input:0 ~expected:(Some "a");
+      test_find ~input:1 ~expected:(Some "b");
+      test_find ~input:2 ~expected:(Some "j");
```
[the remark this answers is not in the document]
Comment (maxmwang) [reply tests/test-units/analysis/overrides_interval_tree_test.ml]: 1. Clarified what I meant by "first" - it returns the first override provided in the list when `of_alist` is called. In the use case of ppx overrides, there shouldn't be a case where there are duplicates, but it is technically allowed. 
2. As discussed in the row, since endpoints are inclusive, we will allow zero width intervals.

File: src/analysis/overrides_interval_tree.mli [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  val create : low:int -> high:int -> payload:'a -> ('a t, string) result
+end
+
+type 'a t
+
+(** Find the tightest interval that contains a given integer point. Runs in O(logn + mlogm)
+    where m is the number of intervals containing the point.
+
+    [find] assumes a total ordering. Ties imply equivalence, and we return the
+    first. *)
```
[the remark this answers is not in the document]
Comment (maxmwang) [reply src/analysis/overrides_interval_tree.mli]: Wrote "`find` assumes that an interval is either contained by or contains every other interval."

Rewrote to what you've suggested.

File: src/analysis/overrides_interval_tree.mli [hunk 4]
```diff
@@ -0,0 +1,34 @@
+(** This interval tree is an immutable data structure that stores mappings from integer
+    intervals to values ['a] and allows efficient queries for intervals that contain a
+    given point.
+
+    Interval tree assumes that input intervals have a total ordering, such as AST node
+    locations.
```
[the remark this answers is not in the document]
Comment (maxmwang) [reply src/analysis/overrides_interval_tree.mli]: Agreed, removed

File: tests/test-dirs/overrides.t (hunk 1 above)
[the remark this answers is not in the document]
Comment (liam923) [reply tests/test-dirs/overrides.t:376]: Now that we switched to being inclusive, is the original back to passing? I'm a bit concerned if it's not.

[the remark this answers is not in the document]
Comment (maxmwang) [reply tests/test-dirs/overrides.t]: Original is passing, reverted.

File: src/analysis/overrides_interval_tree.mli [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    they are not necessary, but are possibly easy to include.
+
+    The general design of the data structure is on
+    {{:https://en.wikipedia.org/wiki/Interval_tree#Centered_interval_tree}this wiki page}. *)
+
+(** [Interval] contains an interval tree entry's range and payload. *)
+module Interval : sig
+  type 'a t
+
+  (** [low] is included in the range. [high] is excluded from the range. Returns [Error] if
```
Comment (liam923) [inline src/analysis/overrides_interval_tree.mli:16]: This comment is out of date now

**Review by liam923 — CHANGES_REQUESTED** [review]

**Comment by liam923** [comment]
I'll hit merge once CI passes. Poke me if I forget (I probably will)
